base/broadcast.jl

```diff
@@ -228,19 +247,46 @@ function broadcast_t(f, ::Type{Any}, As...)
     keeps, Idefaults = map_newindexer(shape, As)
     st = start(iter)
     I, st = next(iter, st)
-    val = f([ As[i][newindex(I, keeps[i], Idefaults[i])] for i=1:nargs ]...)
+    val = f(( _broadcast_getindex(As[i], newindex(I, keeps[i], Idefaults[i])) for i=1:nargs )...)
     B = similar(Array{typeof(val)}, shape)
     B[I] = val
     return _broadcast!(f, B, keeps, Idefaults, As, Val{nargs}, iter, st, 1)
 end
 
-@inline broadcast_t(f, T, As...) = broadcast!(f, similar(Array{T}, broadcast_shape(As...)), As...)
+@inline broadcast_t(f, T, As...) = broadcast!(f, similar(Array{T}, broadcast_indices(As...)), As...)
+
+@generated function broadcast_tup{AT,nargs,n}(f, As::AT, ::Type{Val{nargs}}, ::Type{Val{n}})
+    quote
+        ntuple(n -> (@ncall $nargs f i->_broadcast_getindex(As[i], n)), Val{n})
+    end
+end
+
+function broadcast_c(f, ::Type{Tuple}, As...)
+    shape = broadcast_indices(As...)
+    check_broadcast_indices(shape, As...)
+    n = length(shape[1])
+    nargs = length(As)
+    return broadcast_tup(f, As, Val{nargs}, Val{n})
+end
+broadcast_c(f, ::Type{Any}, a...) = f(a...)
+broadcast_c(f, ::Type{Array}, As...) = broadcast_t(f, promote_eltype_op(f, As...), As...)
 
 """
     broadcast(f, As...)
 
-Broadcasts the arrays `As` to a common size by expanding singleton dimensions, and returns
-an array of the results `f(as...)` for each position.
+Broadcasts the arrays, tuples and/or scalars `As` to a container of the
+appropriate type and dimensions. In this context, anything that is not a
+subtype of `AbstractArray` or `Tuple` is considered a scalar. The resulting
+container is stablished by the following rules:
+
+- If all the arguments are scalars, it returns a scalar.
+- If the arguments are tuples and zero or more scalars, it returns a tuple.
+- If there is at least an array in the arguments, it returns an array
+  (and treats tuples as 1-dimensional arrays) expanding singleton dimensions.
+
+A special syntax exists for broadcasting: `f.(args...)` is equivalent to
+`broadcast(f, args...)`, and nested `f.(g.(args...))` calls are fused into a
+single broadcast loop.
```

Comment:
You need to run `./julia doc/genstdlib.jl` to update the manual with this docstring.

Also, there is a manual section on broadcasting, and it would be good to include an example acting on strings or similar, e.g. `string.(["one","two","three","four"],':',1:4)`.